File: sdk/go/provider/dropmail_click.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"

	http "github.com/bogdanfinn/fhttp"
)

// dropmail.click：独立后端临时邮箱（MX 自研 mail.dropmail.click，与 dropmail.me 无关）
// 免注册、无鉴权、纯 REST API；邮箱有效期 10 分钟。

const dropmailClickBaseURL = "https://dropmail.click"

// DropmailClickGenerate 创建 dropmail.click 临时邮箱
// POST /api/v1/public/mailbox → {address, created_at, expires_at}
func DropmailClickGenerate() (*CreatedMailbox, error) {
	client := HTTPClient()
	req, err := http.NewRequest("POST", dropmailClickBaseURL+"/api/v1/public/mailbox", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "dropmail-click generate"); err != nil {
		return nil, err
	}
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data struct {
		Address   string `json:"address"`
		ExpiresAt string `json:"expires_at"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, fmt.Errorf("dropmail-click: 无效响应 %s", string(raw))
	}
	if data.Address == "" {
		return nil, fmt.Errorf("dropmail-click: 缺少 address 字段")
	}
	return &CreatedMailbox{
		Channel:   "dropmail-click",
		Email:     data.Address,
		Token:     data.Address,
		ExpiresAt: data.ExpiresAt,
	}, nil
}

// DropmailClickGetEmails 获取 dropmail.click 邮件列表
// GET /api/v1/public/mailbox/{email} → {messages:[{id, address, from, subject, text, html, ...}]}
func DropmailClickGetEmails(email string) ([]NormEmail, error) {
	if email == "" {
		return nil, fmt.Errorf("dropmail-click: 缺少邮箱地址")
	}
	client := HTTPClient()
	reqURL := fmt.Sprintf("%s/api/v1/public/mailbox/%s", dropmailClickBaseURL, url.PathEscape(email))
	req, err := http.NewRequest("GET", reqURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", getCurrentUA())
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "dropmail-click getEmails"); err != nil {
		return nil, err
	}
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data struct {
		Messages []json.RawMessage `json:"messages"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil, err
	}
	if len(data.Messages) == 0 {
		return []NormEmail{}, nil
	}
	return NormalizeRawMessages(data.Messages, email)
}

File: sdk/npm/src/providers/dropmail-click.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "dropmail-click";
const BASE_URL = "https://dropmail.click";

const UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/149.0.0.0 Safari/537.36";

/**
 * 创建 dropmail.click 临时邮箱
 * POST /api/v1/public/mailbox → { address, created_at, expires_at }
 * 邮箱有效期 10 分钟；token = email
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const resp = await fetchWithTimeout(`${BASE_URL}/api/v1/public/mailbox`, {
    method: "POST",
    headers: { "User-Agent": UA, Accept: "application/json" },
  });
  if (!resp.ok) {
    throw new Error(`dropmail-click: 创建邮箱失败 ${resp.status}`);
  }
  const data = (await resp.json()) as { address?: string; expires_at?: string };
  if (!data.address) {
    throw new Error("dropmail-click: 无效响应，缺少 address");
  }
  return {
    channel: CHANNEL,
    email: data.address,
    token: data.address,
    expiresAt: data.expires_at,
  };
}

/**
 * 获取 dropmail.click 邮件列表
 * GET /api/v1/public/mailbox/{email} → { messages: [{id, address, from, subject, text, html}] }
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const addr = token || email;
  if (!addr) {
    throw new Error("dropmail-click: 缺少邮箱地址");
  }
  const resp = await fetchWithTimeout(
    `${BASE_URL}/api/v1/public/mailbox/${encodeURIComponent(addr)}`,
    {
      method: "GET",
      headers: { "User-Agent": UA, Accept: "application/json" },
    },
  );
  if (!resp.ok) {
    throw new Error(`dropmail-click: 获取邮件失败 ${resp.status}`);
  }
  const data = (await resp.json()) as DropmailClickResponse;
  if (!Array.isArray(data.messages)) {
    return [];
  }
  return data.messages.map((msg) =>
    normalizeEmail(
      {
        id: String(msg.id ?? ""),
        from: msg.from || "",
        to: msg.address || addr,
        subject: msg.subject || "",
        text: msg.text || "",
        html: msg.html || "",
        date: msg.received_at || msg.date || "",
        isRead: false,
        attachments: [],
      },
      addr,
    ),
  );
}

type DropmailClickMessage = {
  id?: string;
  address?: string;
  from?: string;
  subject?: string;
  text?: string;
  html?: string;
  raw?: string;
  received_at?: string;
  date?: string;
};

type DropmailClickResponse = {
  messages?: DropmailClickMessage[];
};
